src/profile-logic/call-tree.js

```diff
@@ -549,7 +551,383 @@ export function getCallTree(
+  ensureExists(summaryPerFunctions[0]);
+  if (summaryPerFunctions.length === 1) {
+    return summaryPerFunctions[0];
+  }
+  const callNodeSelf = new Float32Array(summaryPerFunctions[0].callNodeSelf);
+  const callNodeTotal = new Float32Array(summaryPerFunctions[0].callNodeTotal);
+  for (let i = 1; i < summaryPerFunctions.length; i++) {
+    const { callNodeSelf: self, callNodeTotal: total } = summaryPerFunctions[i];
+    for (let j = 0; j < self.length; j++) {
+      callNodeSelf[j] += self[j];
+      callNodeTotal[j] += total[j];
+    }
+  }
+  return { callNodeSelf, callNodeTotal };
+}
+
+function _subtractSummaryPerFunctions(
+  initial: _SummaryPerFunction,
+  subtracted: _SummaryPerFunction
+) {
+  const callNodeSelf = new Float32Array(initial.callNodeSelf);
+  const callNodeTotal = new Float32Array(initial.callNodeTotal);
+  for (let j = 0; j < callNodeSelf.length; j++) {
+    callNodeSelf[j] -= subtracted.callNodeSelf[j];
+    callNodeTotal[j] -= subtracted.callNodeTotal[j];
+  }
+  return { callNodeSelf, callNodeTotal };
+}
+
+/**
+ * The following code belongs to an optimization that improves the performance of the call tree
+ * significantly, by caching lots of computed data.
+ * It was implemented in https://github.com/firefox-devtools/profiler/pull/4227
+ */
+
+/** ensures that the slices are not to large for any reasonably sized sample */
+const MAX_SAMPLE_SLICE_NUMBER = 300;
+const MAX_SLICES_PER_LEVEL = 5;
+
+type SummarySpecificFuncs<T> = {|
+  /** compute the summary information for the sample range (the end index is not included) */
+  compute: (startIndex: number, endIndex: number) => T,
+  /** combine several instances to one */
+  combine: (summaryPerFunctions: T[]) => T,
+  /** subtract second instance from the first */
+  subtract: (initial: T, subtracted: T) => T,
+|};
+
+/**
+ * Node in the multilevel cache for the function list information
+ * (a mapping of functions to the number of stack samples they appear in) over time:
+ *
+ * The whole sample range is split into slices of equal size.
+ * We store for every slice the function list information.
+ * The slices are further split recursively into smaller, non overlapping,
+ * slides until the slices contain at mosty MAX_SAMPLE_SLICE_NUMBER samples.
+ *
+ * Now when we want to get the function list information for a specific range,
+ * we combine the information from all tree nodes on the root level
+ * that are fully contained in the range. We use the sub slices to recursively
+ * add the information for the borders.
+ * This way we only compute new information for at most 2 * (MAX_SLICES_PER_LEVEL - 1) slices.
+ * Which speeds up the computation.
+ *
+ * The information in the tree nodes themselfes is only computed when needed,
+ * and then cached.
+ *
+ * @param T The type of the information that is stored in the tree nodes,
+ *          e.g. the function list information (_SummaryPerFunction).
+ */
+class SummaryCacheTreeNode<T> {
+  _startIndex: number;
```

Comment:
Please say that this is an index into the samples. Specifically, into the samples of the filtered thread, not the preview filtered thread. Is that right?

Reply:
Sounds right.